### PR TITLE
Generalize the Convolution Depth of Encoders

### PR DESCRIPTION
**Pull request by GabrielAlacchi**
My team's ablation study of this paper for the Neurips Reproducibility Challenge requires us to experiment with different encoder architectures, including the ImpalaCNN that was already added to the codebase. This change allows for flexibility in the depth of the convolutional layer of the encoder. I achieve this by:
1. Add a `local_layer_depth` property to each encoder
2. Modifying the various SpatioTemporal trainers to initialize their
classifiers using this property as opposed to the hardcoded 128.

Additionally, I also included the fmaps keyword argument to ImpalaCNN's `forward` method using the 3rd or 4th block (depending on whether we want to downsample or not) output as the convolutional layer for computing the global-local loss.

For the time being, my team is working from this repository https://github.com/GabrielAlacchi/atari-ari-extensions where we've extended some of these classes to achieve this behavior, however I figured I'd share this change with you as well if you're interested!

### CODE REVIEW

File: atariari/methods/encoders.py [hunk 1]
```diff
@@ -124,6 +136,10 @@ def __init__(self, input_channels, args):
             )
         self.train()
 
+    @property
+    def convolution_depth(self):
+        return self.main[4].out_channels
```
Comment (ankeshanand) [inline atariari/methods/encoders.py]: The name `convolution_depth` might be a bit vague, since it doesn't tell what is the layer we are referring to. Might be better to call the property `depths` and then call `self.encoder.depths[-2]`
```
def depths(self):
    return self._depths
```

File: atariari/methods/global_local_infonce.py [hunk 2]
```diff
@@ -35,9 +35,9 @@ def __init__(self, encoder, config, device=torch.device('cpu'), wandb=None):
         self.device = device
         if self.use_multiple_predictors:
             # todo remove the hard coded 11x8
-            self.classifiers = [nn.Linear(self.encoder.hidden_size, 128).to(device) for _ in range(11*8)]
+            self.classifiers = [nn.Linear(self.encoder.hidden_size, self.encoder.convolution_depth).to(device) for _ in range(11*8)]
```
Comment (ankeshanand) [inline atariari/methods/global_local_infonce.py]: `self.encoder.depths[-2]`

File: atariari/methods/global_local_infonce.py [hunk 3]
```diff
@@ -35,9 +35,9 @@ def __init__(self, encoder, config, device=torch.device('cpu'), wandb=None):
...
         else:
-            self.classifier1 = nn.Linear(self.encoder.hidden_size, 128).to(device)
+            self.classifier1 = nn.Linear(self.encoder.hidden_size, self.encoder.convolution_depth).to(device)
```
Comment (ankeshanand) [inline atariari/methods/global_local_infonce.py]: Same for the all the other places

File: atariari/methods/encoders.py [hunk 4]
```diff
@@ -68,16 +68,28 @@ def _make_layer(self, in_channels, depth):
...
-        if self.downsample:
-            out = self.layer3(self.layer2(self.layer1(out)))
-        else:
-            out = self.layer4(self.layer3(self.layer2(self.layer1(out))))
-        return F.relu(self.final_linear(self.flatten(out)))
+    @property
+    def convolution_depth(self):
+        return self.depths[-1]
+
+    def forward(self, inputs, fmaps=False):
```
Comment (ankeshanand) [inline atariari/methods/encoders.py]: I assume these are the changes you needed to make ImpalaCNN work?

File: atariari/methods/encoders.py (hunk 1 above)
Comment (GabrielAlacchi) [reply atariari/methods/encoders.py]: From a theoretical standpoint what's the benefit of using the 2nd to last depth as the input for the Local-Local and Global-Local score functions? In our experiments we are using the final layer for ImpalaCNN.

File: atariari/methods/encoders.py (hunk 4 above)
Comment (GabrielAlacchi) [reply atariari/methods/encoders.py]: Yes we needed to pass a convolutional layer as 'f5' to the InfoNCE trainers.

File: atariari/methods/encoders.py (hunk 1 above)
Comment (ankeshanand) [reply atariari/methods/encoders.py]: We want to ensure the receptive field for each patch in the feature map is limited (using the second last layer in NatureCNN limits the receptive field to 1/16th of the full observation). This means that loss is distributed across all the regions of the input, and the encoder can't resort to using only one predictable feature. 

You might want to calculate the receptive field for different layers in the Impala architecture and use the appropriate layer.

Comment (GabrielAlacchi) [reply atariari/methods/encoders.py]: This depths[-2] is something we may want to vary depending on our encoder architecture. What do you think about renaming this to something like `intermediate_layer_depth`?

Comment (ankeshanand) [reply atariari/methods/encoders.py]: I see, yes using a fixed layer would be problematic. I think your original solution makes sense then, how about renaming it to `local_layer_depth`?

Comment (GabrielAlacchi) [reply atariari/methods/encoders.py]: I like that name. I made the change!
